File: datastructure&algorithm/ds_algo_java/src/algorithms/dfs/bj_13023_2/Main.java
package algorithms.dfs.bj_13023_2;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static ArrayList<Integer>[] map;
    static int n,m;
    static boolean flag = false;

    public static void main(String[] args) throws IOException{
        String[] nm = br.readLine().split(" ");
        n = Integer.parseInt(nm[0]);
        m = Integer.parseInt(nm[1]);

        map = new ArrayList[n];
        for(int i=0;i<n;i++){
            map[i] = new ArrayList<Integer>();
        }

        for(int i=0;i<m;i++){
            String[] se = br.readLine().split(" ");
            int s = Integer.parseInt(se[0]);
            int e = Integer.parseInt(se[1]);
            map[s].add(e);
            map[e].add(s);
        }

        for(int i=0;i<n;i++){
            boolean[] visited = new boolean[n];
            visited[i] = true;
            dfs(i,0,visited);
            if(flag) break;
        }
        System.out.println(flag?1:0);
    }
    static void dfs(int start,int depth,boolean[] visited){
        if(depth==4){//있음
            flag = true;
            return;
        }

        for(int i:map[start]){
            if(visited[i]) continue;

            visited[i] = true;
            dfs(i,depth+1,visited);
            visited[i] = false;
        }


    }

}

File: datastructure&algorithm/ds_algo_java/src/algorithms/dfs/bj_13023_2/Main.kt
package algorithms.dfs.bj_13023_2;
import java.io.BufferedReader;
import java.io.InputStreamReader;

val br = BufferedReader(InputStreamReader(System.`in`))
var flag = false
fun main(){
    val nm = br.readLine().split(" ")
    val n = nm[0].toInt()
    val m = nm[1].toInt()

    val map = Array<MutableList<Int>>(n){ mutableListOf() }

    repeat(m){
        val se = br.readLine().split(" ")
        val s = se[0].toInt()
        val e = se[1].toInt()

        map[s].add(e)
        map[e].add(s)
    }

    for(i in 0..n-1){
        val visited = Array<Boolean>(n){false}
        visited[i] = true
        dfs(i,1,visited,map)
        if(flag) break
    }

    val lambda = {f:Boolean->
        when(f){
            true->1
            false->0
        }
    }

    println(lambda(flag))

}

fun dfs(node:Int, friends:Int,visited:Array<Boolean>,map:Array<MutableList<Int>>){
    if(friends==5){
        flag = true
        return
    }

    for(i in map[node]){
        if(visited[i]) continue

        visited[i] = true
        dfs(i,friends+1,visited,map)
        visited[i] = false
    }
}